service/importer/service_file.go

```diff
@@ -4,70 +4,84 @@ import (
 
 // validateServiceFile returns a list of warnings.
-func validateServiceFile(data []byte) ([]string, error) {
-	var body interface{}
-	if err := yaml.Unmarshal(data, &body); err != nil {
-		return nil, fmt.Errorf("parse mesg.yml error: %s", err)
-	}
-	body = convert(body)
-	result, err := validateServiceFileSchema(body, "service/importer/assets/schema.json")
-	if err != nil {
-		return nil, err
-	}
-	var warnings []string
-	if !result.Valid() {
-		for _, warning := range result.Errors() {
-			warnings = append(warnings, warning.String())
+func validateServiceFile(data []byte) []string {
+	var service ServiceDefinition
+	if err := yaml.UnmarshalStrict(data, &service); err != nil {
+		errs, ok := err.(*yaml.TypeError)
+		if !ok {
+			return []string{err.Error()}
 		}
+		return errs.Errors
 	}
-	if depKeyWarning := validateServiceFileDependencyKey(body); depKeyWarning != "" {
-		warnings = append(warnings, depKeyWarning)
-	}
-	return warnings, nil
+	return validateServiceStruct(&service)
 }
 
-func validateServiceFileDependencyKey(data interface{}) (warning string) {
-	s, _ := data.(map[string]interface{})
-	dep, _ := s["dependencies"].(map[string]interface{})
-	if dep[ConfigurationDependencyKey] != nil {
-		return fmt.Sprintf("cannot use %q as dependency key", ConfigurationDependencyKey)
-	}
-	return ""
-}
+func validateServiceStruct(service *ServiceDefinition) []string {
+	en := en.New()
+	uni := ut.New(en, en)
+	trans, _ := uni.GetTranslator("en")
+	validate := validator.New()
+	validate.RegisterValidation("port", xvalidator.IsPort)
+	validate.RegisterTranslation("port", trans, func(ut ut.Translator) error {
+		return ut.Add("port", "{0} must be a valid port", false)
+	}, func(ut ut.Translator, fe validator.FieldError) string {
+		t, _ := ut.T("port", fe.Field(), namespacePrefix)
+		return t
+	})
+	validate.RegisterValidation("domain", xvalidator.IsDomainName)
+	validate.RegisterTranslation("domain", trans, func(ut ut.Translator) error {
+		return ut.Add("domain", "{0} must respect domain-style notation. Eg: author.name", false)
+	}, func(ut ut.Translator, fe validator.FieldError) string {
+		t, _ := ut.T("domain", fe.Field())
+		return t
+	})
+	en_translations.RegisterDefaultTranslations(validate, trans)
 
-func validateServiceFileSchema(data interface{}, schemaPath string) (*gojsonschema.Result, error) {
-	schemaData, err := assets.Asset(schemaPath)
+	err := validate.Struct(service)
+	warnings := []string{}
 	if err != nil {
-		return nil, err
+		errs := err.(validator.ValidationErrors)
+		for _, e := range errs {
+			// Remove the name of the struct and the field from namespace
+			trimmedNamespace := strings.TrimPrefix(e.Namespace(), namespacePrefix)
+			trimmedNamespace = strings.TrimSuffix(trimmedNamespace, e.Field())
+			// Only use it when in-cascade field
+			namespace := ""
+			if e.Field() != trimmedNamespace {
+				namespace = trimmedNamespace
+			}
+			warnings = append(warnings, fmt.Sprintf("%s%s", namespace, e.Translate(trans)))
```

Comment:
I change from custom error output to use the `Translate` function available in govalidator that format nicely the error.

I add namespace to have specific "path" to the error, eg:
```
        * Tasks[decodeLog].Inputs[logIndex].Type must be one of [String Number Boolean Object Any]
```
instead of default:
```
        * Type must be one of [String Number Boolean Object Any]
```

Reply:
Nice! ❤️ 

Is would be nice to have fields in loware case (like we do in mesg.yaml) so 

`task[XXX].inputs[YYY].type bla bla bla`

Reply:
Yes. But I don't know what is it uppercased.
For instance, `e.Field()` is containing the field declare in the json label, as opposite as `StructField` that return the Field name of the Go struct.

Reply:
I mean e.Fields because all fields in mesg.yml are lowerace

Reply:
@krhubert the modif you did, didn't work.
The output is:
```
        * NameName is a required field
```

`e.Translate(trans)` is returning `Type must be one of [String Number Boolean Object Any]` (uppercased).
The system I did is only adding the beginning of the sentence. I don't want to spend hours fixing this problem. It's fine ;)

Reply:
If you try to lowercase all the string, then the user's camelcase key will also be lowercase. I don't see easy solution for this very small problem.